src/backends/policy_engine/souffle/datalog_lowering_visitor.cc

```diff
@@ -0,0 +1,103 @@
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+
+#include "src/backends/policy_engine/souffle/datalog_lowering_visitor.h"
+
+#include "src/common/logging/logging.h"
+#include "src/ir/attributes/attribute.h"
+#include "src/ir/attributes/int_attribute.h"
+#include "src/ir/attributes/string_attribute.h"
+#include "src/ir/datalog/operation.h"
+#include "src/ir/module.h"
+#include "src/ir/operator.h"
+
+namespace raksha::backends::policy_engine::souffle {
+
+using DatalogOperation = ir::datalog::Operation;
+using DatalogSymbol = ir::datalog::Symbol;
+using DatalogNumber = ir::datalog::Number;
+using DatalogAttributeList = ir::datalog::AttributeList;
+using DatalogOperandList = ir::datalog::OperandList;
+using DatalogIsOperationFact = ir::datalog::IsOperationFact;
+using DatalogAttribute = ir::datalog::Attribute;
+using DatalogAttributePayload = ir::datalog::AttributePayload;
+using DatalogStringAttributePayload = ir::datalog::StringAttributePayload;
+using DatalogNumberAttributePayload = ir::datalog::NumberAttributePayload;
+using IrOperation = ir::Operation;
+
+static DatalogAttributePayload GetPayloadForAttribute(ir::Attribute attr,
+                                                      ir::SsaNames &ssa_names) {
+  if (auto int_attr = attr.GetIf<ir::Int64Attribute>()) {
+    return DatalogNumberAttributePayload(DatalogNumber(int_attr->value()));
+  } else if (auto string_attr = attr.GetIf<ir::StringAttribute>()) {
+    return DatalogStringAttributePayload(DatalogSymbol(string_attr->value()));
+  } else {
+    LOG(FATAL) << "Unknown attribute kind.";
+  }
+  // Unreachable, just to placate compiler.
+  return DatalogStringAttributePayload(DatalogSymbol(""));
+}
+
+static DatalogAttribute TranslateIrNodeToDatalog(
+    std::pair<std::string, ir::Attribute> name_attr_pair,
+    ir::SsaNames &ssa_names) {
+  return DatalogAttribute(
+      DatalogSymbol(std::move(name_attr_pair.first)),
+      GetPayloadForAttribute(std::move(name_attr_pair.second), ssa_names));
+}
+
+static DatalogSymbol TranslateIrNodeToDatalog(ir::Value value,
+                                              ir::SsaNames &ssa_names) {
+  return DatalogSymbol(value.ToString(ssa_names));
+}
+
+// Turn a range of IR nodes in a C++ container into a Datalog-style linked
+// list.
+template <class ListT, class IrNodeIter>
+static ListT RangeToDatalogList(IrNodeIter ir_node_list_begin,
+                                IrNodeIter ir_node_list_end,
+                                ir::SsaNames &ssa_names) {
```

Comment:
Instead of passing in a range, the following provides a more flexible API: 
```suggestion
template <class ListT, class T>
static ListT IrNodeToDatalogList(T ir_nodes, ir::SsaNames &ssa_names) {
  
```

Reply:
I think this setup would be hard to use with recursive traversal of the list, which is the most natural way to do this because of the structure of a Datalog list type.

Reply:
This could also be made more natural by integrating #539 .

Reply:
Fixed with use of `fold`.